### PR TITLE
Refactor

### LINKED ISSUE #53: Doesn't work in diff editor view
```viewColumn``` is undefined. Possible workaround is to create separate window for undefined view columns

### LINKED ISSUE #78: Does :registers work?
Thanks for the awesome plugin--it's way better than the others. 

Should `:registers` work? It doesn't seem to do anything for me.

### LINKED ISSUE #100: Incorrect handling of tab indentation with rectangular selection
Rectangular selections using Ctrl-V become non-rectangular, as tabs shift part of the selection:

Screenshot_2019-12-17_13-00-59

### LINKED ISSUE #152: Cursor position desync in tab-indented Python files
For .py files, NeoVim defaults to tabstop=8. This completely breaks the cursor positioning in vscode-neovim for tab-indented files.

This can be reproduced with a simple file with .py extension without modeline. My init.vim does not contain any settings related to indentation.

- VSCode 1.41.1-1576681836
- NeoVim 0.5.0+ubuntu1+git202001122018-05ea3c1-00e710e-8b2f2dc~ubuntu18.04.1
- vscode-neovim 0.0.49

### LINKED ISSUE #155: s/foo/bar inserts tab after replacement

### LINKED ISSUE #167: Does not play well with editorconfig or detected indentation
It feels like neovim inherits the indentation settings once (when neovim gets called for the first time).
When the indentation changes later on (due to detected indentation settings or through editorconfig), neovim isn't notified of these changes anymore and continues using the old indentation settings.

This leads to weird behaviour when shifting text around with `<` and `>` (and derivatives, including `<<` and `>>`) if
- VSCode's default indentation is spaces, but it detects a file being indented with tabs (and vice versa)
- VSCode's default indentation is spaces, but editorconfig wants a file to be indented with tabs (and vice versa)

I haven't tested this, but I imagine things getting weird as well if editorconfig only changes the indentation width away from VSCode's default.

I thought I could work around this issue by adding the editorconfig to my neovim plugins, but those settings are overridden as well by VSCode calling neovim.

### LINKED ISSUE #187: Can't move  beyond the 500th character of a line
I don't why I can't move the cursor beyond  the 500th character by using j or w.  It works fine by clicking the mouse.

### LINKED ISSUE #202: Make :messages work
I'm trying to troubleshoot a command I'm defining in nvim to send commands to VSCode. It's hard to do because I can't view the messages buffer.

### LINKED ISSUE #207: Recording Macro does some weird stuff
Steps to reproduce problem:

1. Start recording macro, e.g. `qq`
2. Open new line to edit below, e.g. `o`
3. Type something

The text will be put on the line above instead of the original line and the cursor will be before the character being input instead of after. If you continue moving around, going in and out of insert mode, and typing in text, it will sometimes put in copies of lines from down below. If you try to undo, it will start duplicating or deleting parts of lines randomly (I haven't found the pattern yet). Seems like another insert mode problem.

I have nothing in my init.vim file and have disabled all other vs code plugins.

### LINKED ISSUE #220: Normal mode movements break with multiple panes (3 or more)
When three or more panes are opened in such a configuration...

```
+----+----+
|    |    |
+----|    |
|    |    |
+----+----+
```

...normal mode movements become unresponsive in all but the most top left pane.

The cursor also does not work when trying to click around the document to navigate.

However, going into insert mode and navigating using the arrow keys still works in all panes.

### LINKED ISSUE #223: Can't get the plugin working: TextEditor(vs.editor.ICodeEditor:2,$model11) has been disposed
Hey!

Have installed neovim and that works, but installed vscode-neovim does not. When I open vscode with the neovim path specific and the plugin installed, it opens a bunch of binary files (see screenshot) and display an error:

image

```
vscode-neovim: Error applying neovim edits, please report a bug, error: TextEditor(vs.editor.ICodeEditor:2,$model11) has been disposed
```

Two of the other files it opens are the vim reference manual, twice.

Then, none of the navigation commands such as `w`, `hjkl` work - although the ex commands do.

Any ideas what the problem is / how to resolve this?

### LINKED ISSUE #226: Cursor freezes after "search->Open in editor->ctrl-click to open"
What I did:
1. use vscode's native search and open results in editor:

2. in the results, ctrl-click the file name to open the file:

Then I'm in the file but I can no longer see the cursor moving with vim keys such as "j" "k", "gg".
However, the cursor does move under the hood - because "jdd" can still delete the line after the current line. But I can no longer see it moving with "j".

### LINKED ISSUE #239: cursor is in the wrong place in tab-indent file
1. 4-space-indent file,ujhk move cursor right.
2. after change all space to tab,ujhk movement is chaos,can't reach the head of line with ^ or h,can't search the right words under the cursor, etc...

### LINKED ISSUE #264: tabs indentation
There's an issue with the cursor width when the file has tabs for indentation, where the cursor is as wide as the tab stop, no problems when using spaces for indentation. Is this a known issue or someone can corroborate?

Thank you.

### LINKED ISSUE #288: bug: arrows don't work for cmdline history navigation
I use regexen extensively in vim / neovim. Often I will mess up a regex somehow and need to correct it, and I do that by editing it from my search history (i.e. in vim, hitting "/" then pressing up arrow until I reach the regex I want to edit). How do I do this in vscode-vim? I've been having to spawn vim in a terminal, do the regex stuff, and then exit.

### LINKED ISSUE #289: In navigate mode, random cursor motion
Hello,

I've been having issues with navigation mode, especially for softwrapped paragraphs. I've made changes to my .vimrc (which I source) per #259 so that I can do that. Nav-mode (especially gk/gj) seems to randomly displace my cursor, and gets it caught in weird cycles; for example trying to go down a softwrapped line can result in going down several lines, then cycling back to the top of the paragraph before getting to the softwrapped line to-be-edited. Clicking will insert the cursor in a random spot somewhere in the document, often selecting a large swath of random text in the process. Interestingly, none of this happens in insert mode. It seems to go away sometimes if I close then reopen the entire program. It somehow seems related to whether I've clicked onto a tab in the editor view panel, clicked away (to some other tab) then clicked back to that tab. If I do that, it often makes these issues go away.

Obviously I really want to get this sorted, but I'm not sure what other info would be helpful for troubleshooting. I'm happy to provide whatever information is deemed necessary.

Thanks for the great plugin.

### LINKED ISSUE #296: For ex-mode g/pattern/p, the p is not working
On OSX 10.15.5
with VSC 1.46.1 + Neo Vim v0.0.50
with NeoVim Path set to /usr/local/bin/nvim
and nvim --version is NVIM v0.4.3

Suppose that I search for all lines with _xyz_ in them and print them with

> :g/xyz/p

The print result shows nothing although the cursor has been moved to the last _xyz_. Also, when I am typing the p, VSC is trying to helpfully autocomplete that as packadd packloadall ...

If I delete all lines with _xyz_ in them

> :g/xyz/d

So the delete in fact works.

Basically, the print results just aren't displaying in some sort of popup.

### LINKED ISSUE #300: Search and Replace does not work with angle brackets
Consider the text and command:
```
<xml>
```
`s/\v\<xml/<lmx/`

We should expect that the text is changed from xml to lmx.

What actually happens: entire match is deleted as if the replacement pattern was empty.

Neo Vim plugin: 0.0.50
Neovim: 0.4.2
OS: Windows

### LINKED ISSUE #312: Line desync on line deletions
Sometimes I expected desync with `dd`.

**Steps to reproduce**

Tested on [this file](https://github.com/Shatur95/akd/blob/893cf5db597fba3565aad05c40f82164e41f719d/src/layout.cpp#L33) with clean config.

```vim
dd
gk
u
gk
dd
u
gj
dd
```

**Current behavior**

After the last `dd`, the line above the cursor will be deleted.

**Expected behavior**

After the last `dd`, the current line will be deleted.

### LINKED ISSUE #331: Splitting using shortcuts is broken in 0.0.52
`C-w v` and `C-w s` work as expected on 0.0.50 and do nothing on 0.0.52.

`:sp` and `:vsp` still work as expected. Other `C-w [key]` combinations does not seem to be affected - e.g. `C-w o` , `C-w {hjkl}` , `C-w >` still work.

### PR DESCRIPTION
**Pull request by asvetliakov**
Refactor/rewrite almost all code. Simplicities and fixes many things related to buffer/window/cursor synchronisation

fixes #53 , fixes #187 , fixes #220 , fixes #223, fixes #202, fixes #78, fixes #226 

Fixes tab related problems, fixes: #275, fixes #239, fixes #264, fixes #167, fixes #100, fixes #152 , fixes #289 

Fixes macro recording problems: fixes #207 

Fixes some command line problems: fixes #155, fixes #288

Fixes some desync issues: fixes #312 

Ohter fixes: fixes #331 , fixes #300 , fixes #296

### CODE REVIEW

**Comment by theol0403** [comment]
This is amazing, I've already switched to this branch and it has solved pretty much every issue I've had with this plugin (other than command mode)! Thanks a lot for making this :smiley: 

I have just noticed one regression, and it is that the filetype does not seem to be properly set. For example, this causes the vim-commentary plugin to comment using `/* */` instead of `#` in a python file.

**Comment by asvetliakov** [comment]
@theol0403 
>Edit: Another thing I've noticed is that vscode no longer updates the buffer when there is an external file modification.

I.e. when file is modified outside of vscode? Can you provide repro steps because [this test](https://github.com/asvetliakov/vscode-neovim/blob/5b1581ec2cadcd6dec03f2e7957f35ec8cfe4df6/src/test/suite/external-changes.test.ts#L17) passes ok

**Comment by Shatur95** [comment]
> I have just noticed one regression, and it is that the filetype does not seem to be properly set. For example, this causes the vim-commentary plugin to comment using /* */ instead of # in a python file.

This is an interesting notice and could be fixed. But maybe you find this mappings useful:

```vim
xmap gc <Plug>VSCodeCommentary
nmap gc <Plug>VSCodeCommentary
omap gc <Plug>VSCodeCommentary
nmap gcc <Plug>VSCodeCommentaryLine
```

File: vim/vscode-options.vim [hunk 1]
```diff
@@ -52,6 +52,9 @@ set foldmethod=manual
 set autoindent
 set smartindent
 
+" split/nosplit doesn't work currently, see https://github.com/asvetliakov/vscode-neovim/issues/329
+set inccommand=""
```
Comment (Shatur95) [inline vim/vscode-options.vim:56]: You probably want the following:
```suggestion
set inccommand=
```

**Comment by theol0403** [comment]
> I.e. when file is modified outside of vscode? Can you provide repro steps because [this test](https://github.com/asvetliakov/vscode-neovim/blob/5b1581ec2cadcd6dec03f2e7957f35ec8cfe4df6/src/test/suite/external-changes.test.ts#L17) passes ok

Humm, I actually haven't had the issue since I wrote that. All I noticed was a few times when I made some external file changes and then came back to vscode and tried to save it gave me an "overwrite" warning. Has not happened since, and it does seem to properly handle external changes. Maybe I had an unsaved buffer in VSCode. You can just ignore this.

VSCodeCommentary isn't good enough for me because I like `gcgc` and `dagc` :wink:.

Again, thanks a ton for all these improvements, this is very exciting!

**Comment by asvetliakov** [comment]
@theol0403 
>VSCodeCommentary isn't good enough for me because I like gcgc and dagc

Filetype detection should be fixed in this branch

**Comment by theol0403** [comment]
Sometimes, when I go into insert mode with something that modifies the buffer length (ie change commands) and start typing too fast, the cursor does not update fast enough and it inserts text in the wrong place. 

For example, here I type `ciwadditional` but this happens:

Peek 2020-09-04 20-14


I am on the latest commit of this branch.

**Comment by asvetliakov** [comment]
@theol0403 Interesting, that shouldn't happen and i haven't noticed it personally. May be something with your `init.vim` ? check `autocmd` commands especially if you have `InsertEnter`. Another point can be you accidentally entered macro recording mode - this one is special

**Comment by asvetliakov** [comment]
If someone is using this branch and at same time on Windows, please let me know if it works/doesn't work for you. Win CI build is totally failing and i don't have Win currently to check if it's CI issue or extension issue

**Comment by theol0403** [comment]
Not macro mode, as pressing `q` prompted me for a register. 
I have a large init.vim, but no custom `InsertEnter` autocmds other than what is in plugins. I will disable my init.vim and try to reproduce.

Umm, deleting all the contents of my `init.vim` file causes the cursor to freeze and be unresponsive on VSCode launch.

**Comment by asvetliakov** [comment]
>Umm, deleting all the contents of my init.vim file causes the cursor to freeze and be unresponsive on VSCode launch.

Try reload the window one/few times. Perhaps i've missed some deadlock here (happened to me too)

>About the insert mode issue, I've noticed that the delay for the cursor to update to the new position is about the same time that "clangd: parsing includes" flashes on my statusbar.

Is this coming from vscode extension or from vim? I guess if it happens when entering insert mode it's vim stuff, so try to disable/wrap into `exists('g:vscode')`

**Comment by theol0403** [comment]
Reloading window and restarting VSCode does nothing, still stuck.

It is coming from VSCode extension clangd, and it likely triggered when the buffer is modified (ie when the change command deletes the motion).

**Comment by asvetliakov** [comment]
Try to disable this extension then and see if it still happens. Actually this can be culprit since cursor updates now wait to complete possible pending document changes. But it must set cursor correctly after ¯\_(ツ)_/¯

**Comment by theol0403** [comment]
I disabled the clangd extension and noticed no difference. We haven't ruled out an init.vim issue, I just can't really try disabling it because it renders input useless :stuck_out_tongue:.

**Comment by asvetliakov** [comment]
Try to load console neovim and see if it works with your emptied init.vim

**Comment by theol0403** [comment]
NVIM v0.5.0-dev, and yes console neovim works with empty init.vim.

**Comment by asvetliakov** [comment]
Just tried with commented out/totally empty `init.vim`, works ok. So i'm not sure what is happening here. Check your line ending - Win neovim doesn't like init.vim with LF endings and Linux/Mac neovim doesn't like CRLF

**Comment by theol0403** [comment]
Deleting init.vim, and disabling all extensions in VSCode, works.
The insert mode issue is gone as well.
I will see how adding things affects both issues and get back to you. Luckily the issues are easy to reproduce on my end.

**Comment by theol0403** [comment]
Ok, so VSCode input freezing up happens with empty init.vim but not deleted init.vim.
When init.vim has an empty line, VSCode freezes up. When init.vim has 2 empty lines, VSCode is fine.
Running `dos2unix` on init.vim did change how the empty file looks like in the console (now there are `~` at the beginning of each line) but did not change VSCode behaviour.

**Comment by asvetliakov** [comment]
@theol0403 Are you on win or linux/mac ?

**Comment by theol0403** [comment]
Ubuntu-based linux. I think it might be best for me to take a bit of time to try everything then provide a better report, instead of reporting things as I go.

The latest development is that re-enabling all vscode plugins cause it to freeze again, regardless of extra lines in init.vim. It seems like there is weird timing/loading going on, but I'm not sure.

**Comment by asvetliakov** [comment]
[vscode-neovim-0.0.60.zip](https://github.com/asvetliakov/vscode-neovim/files/5178156/vscode-neovim-0.0.60.zip)
Test version if someone willing to try (rename ext to vsix)

**Comment by theol0403** [comment]
Alright, so I played around more trying to corner the issue.
I'm just going to focus on the issue of neovim not loading properly and leave the cursor jump issue for later.

Unfortunately, it seems like there isn't a single cause for neovim to not load properly. With a 300 line fully-commented out init.vim, and all my vscode extensions enabled, it locks up. With all the vscode extensions disabled, it works. However, by slowly re-enabling extensions, the issue has come back with multiple different plugins. What I mean is that I have about 3 plugins that when they are all disabled, it works fine, but when I enable any one of them (gitlens, wakatime, code-autocomplete), the issue comes back. In fact, when I just enable wakatime, and restart vscode many times, it actually only fails to load neovim *half* of the time. I have also encountered the issue with all those 3 plugins disabled, and also sometimes have not had the issue with all plugins enabled.

So it seems like there is some weird timing/race condition going on. It seems like the longer vscode takes to load, the more likely the lockup is to happen, and the longer init.vim takes to parse, the less likely the lockup is to happen. I'm not sure exactly how to help next.

Some information/observations:
 - Pop!_OS 20.04 LTS x86_64
 - vscode 1.49.0-insiders
 - NVIM v0.5.0-dev
 - vscode-neovim compiled from source from master (as of writing this)
 - everything works fine in console neovim
 - works fine with plugins and init.vim enabled, with vscode-neovim 0.0.52
 - the issue still happens with init.vim completely deleted (as opposed to commented out)

In 0.0.52, the cursor seems to go to block mode on boot (how I can tell neovim has loaded) much later than in 0.0.60. It seems to be ready only after everything has loaded, including semantic highlighting and lsp symbol tree. In 0.0.60, the cursor goes to block mode and is responsive to vim input before everything has finished loading.

Hopefully this gives enough information to pin down the issue, I'm not sure :smile:.
Do you want me to create a new issue for this? It directly relates to refactor but you have already merged and closed this PR :shrug:

**Pull request merged**